src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs

```diff
@@ -288,20 +302,256 @@ private static ProjectItemGroupElement CreateItemGroup(Project project, string f
+        /// target frameworks with their packages
+        /// </summary>
+        /// <param name="project">PR project file</param>
+        /// <param name="userInputFrameworks">List of frameworks</param>
+        /// <param name="assetsFile">Assets file</param>
+        /// <returns></returns>
+        private Dictionary<string, IEnumerable<PRPackage>> GetRequestedPackages(Project project, IList<string> userInputFrameworks, LockFile assetsFile)
+        {
+            var result = new Dictionary<string, IEnumerable<PRPackage>>();
+
+            foreach (var tfm in assetsFile.PackageSpec.TargetFrameworks)
+            {
+                if (userInputFrameworks.Count == 0 || userInputFrameworks.Contains(tfm.ToString())) //If tfm is in the list given by the user or no frameworks were specified
+                {
+                    var projPackages = GetPackageReferencesPerFramework(project, "", tfm.ToString());
+
+                    result.Add(tfm.ToString(), tfm.Dependencies.Select(d => {
+
+                        
+                        string versionStr; //requested version exactly as it is in the proj file
+                        VersionRange version; //requested version value as a range
+                        //If the package is not auto-referenced, get the version from the project file. Otherwise fall back on the assets file
+                        if (!d.AutoReferenced)
+                        {
+                            versionStr = projPackages.Where(p => p.EvaluatedInclude.Equals(d.Name)).First().GetMetadataValue(VERSION_TAG);
+                            version = VersionRange.Parse(versionStr);
+                        }
+                        else
+                        {
+                            versionStr = d.LibraryRange.VersionRange.ToString();
+                            version = d.LibraryRange.VersionRange;
+                        }
+
+                        return new PRPackage {
+                            name = d.Name,
+                            requestedVersion = version,
+                            projectFileRequestedVersion = versionStr,
+                            autoReference = d.AutoReferenced
+                        };
+                    }));
+
+                }
+
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Prepares the dictionary that maps frameworks to packages top-level
+        /// and transitive.
+        /// </summary>
+        /// <param name="userInputFrameworks">A list of framework names</param>
+        /// <param name="assetsFile">Assets file for all targets and libraries</param>
+        /// <param name="requestedVersions">Top-level packages with the requested versions</param>
+        /// <param name="transitive">Include transitive packages in the result</param>
+        /// <returns></returns>
+        private Dictionary<string, Tuple<IEnumerable<PRPackage>, IEnumerable<PRPackage>>> GetResolvedVersions(
```

Comment:
You can just return a List<PRPackage> and have one property there to differentiate between top level vs transitive. That will be more readable then passing two different lists altogether.

Reply:
But later when printing it will take longer time when sorting the object to the 2 lists. I'll need to go through the list again, and figure out which ones are top-level and which ones are transitive, which seems inefficient